dag_in_context/src/dag2svg.rs

```diff
@@ -3,24 +3,69 @@ use std::{
     rc::Rc,
 };
 
-use dot_structures::{Edge, EdgeTy, Graph, Id, Node, NodeId, Stmt, Subgraph, Vertex};
+use dot_structures::{Attribute, Edge, EdgeTy, Graph, Id, Node, NodeId, Stmt, Subgraph, Vertex};
 
-use crate::schema::{Expr, RcExpr, TreeProgram};
+use crate::schema::{Constant, Expr, RcExpr, TreeProgram};
+
+// We need one node for every unique expression, scope pair
+// Expressions may be shared across scopes but should be shown multiple times
+#[derive(Eq, PartialEq, Hash)]
+struct UniqueExpr {
+    expr: *const Expr,
+    scope: *const Expr,
+}
+
+impl UniqueExpr {
+    fn new(scope: *const Expr, expr: &RcExpr) -> Self {
+        UniqueExpr {
+            expr: Rc::as_ptr(expr),
+            scope,
+        }
+    }
+}
 
 struct DotConverter {
-    pub done: HashSet<*const Expr>,
-    pub get_name: HashMap<*const Expr, String>,
+    pub current_scope: *const Expr,
+    pub done: HashSet<UniqueExpr>,
+    pub get_name: HashMap<UniqueExpr, String>,
     pub name_counter: usize,
 }
 
 impl DotConverter {
     pub fn graphviz_id(&mut self, expr: &RcExpr) -> Id {
-        if let Some(name) = self.get_name.get(&Rc::as_ptr(expr)) {
+        if let Some(name) = self
+            .get_name
+            .get(&UniqueExpr::new(self.current_scope, expr))
+        {
             Id::Plain(name.clone())
         } else {
-            let name = format!("{}{}", expr.constructor().name(), self.name_counter);
+            let name = match expr.as_ref() {
+                Expr::DoWhile(_, _) => {
+                    format!("cluster_{}{}", expr.constructor().name(), self.name_counter)
```

Comment:
Why `cluster` instead of `DoWhile`?

Reply:
It's a dot thing: naming it cluster tells dot to create a group of nodes inside a box

Reply:
added comment